mlir/lib/Conversion/QuantumToMQTOpt/QuantumToMQTOpt.cpp

```diff
@@ -0,0 +1,476 @@
+               gateName.compare("ControlledPhaseShift") == 0) {
+      mqtoptOp = rewriter.create<::mqt::ir::opt::POp>(
+          op.getLoc(), outQubitTypes, staticParams, paramsMask, paramsValues,
+          inQubitsValues, inCtrlQubitsValues, inNegCtrlQubitsValues);
+    } else {
+      llvm::errs() << "Unsupported gate: " << gateName << "\n";
+      return failure();
+    }
+
+    // Replace the original with the new operation
+    rewriter.replaceOp(op, mqtoptOp);
+    return success();
+  }
+};
+
+struct QuantumToMQTOpt : impl::QuantumToMQTOptBase<QuantumToMQTOpt> {
+  using QuantumToMQTOptBase::QuantumToMQTOptBase;
+
+  void runOnOperation() override {
+    MLIRContext* context = &getContext();
+    auto* module = getOperation();
+
+    ConversionTarget target(*context);
+    target.addLegalDialect<::mqt::ir::opt::MQTOptDialect>();
+    target.addIllegalDialect<catalyst::quantum::QuantumDialect>();
+
+    // Mark operations legal, that have no equivalent in the target dialect
+    target.addLegalOp<catalyst::quantum::DeviceInitOp>();
+    target.addLegalOp<catalyst::quantum::DeviceReleaseOp>();
+    target.addLegalOp<catalyst::quantum::NamedObsOp>();
+    target.addLegalOp<catalyst::quantum::ExpvalOp>();
+    target.addLegalOp<catalyst::quantum::FinalizeOp>();
+    target.addLegalOp<catalyst::quantum::ComputationalBasisOp>();
+    target.addLegalOp<catalyst::quantum::StateOp>();
+
+    RewritePatternSet patterns(context);
+    QuantumToMQTOptTypeConverter typeConverter(context);
+
+    patterns.add<ConvertQuantumAlloc, ConvertQuantumDealloc,
+                 ConvertQuantumExtract, ConvertQuantumMeasure,
+                 ConvertQuantumInsert, ConvertQuantumCustomOp>(typeConverter,
+                                                               context);
+
+    // Boilerplate code to prevent: unresolved materialization
+    populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
+        patterns, typeConverter);
+    target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      return typeConverter.isSignatureLegal(op.getFunctionType()) &&
+             typeConverter.isLegal(&op.getBody());
+    });
+
+    populateReturnOpTypeConversionPattern(patterns, typeConverter);
+    target.addDynamicallyLegalOp<func::ReturnOp>(
+        [&](func::ReturnOp op) { return typeConverter.isLegal(op); });
+
+    populateCallOpTypeConversionPattern(patterns, typeConverter);
+    target.addDynamicallyLegalOp<func::CallOp>(
+        [&](func::CallOp op) { return typeConverter.isLegal(op); });
+
+    populateBranchOpInterfaceTypeConversionPattern(patterns, typeConverter);
+    target.markUnknownOpDynamicallyLegal([&](Operation* op) {
+      return isNotBranchOpInterfaceOrReturnLikeOp(op) ||
+             isLegalForBranchOpInterfaceTypeConversionPattern(op,
+                                                              typeConverter) ||
+             isLegalForReturnOpTypeConversionPattern(op, typeConverter);
```

Comment:
To be fair, I have no idea what these lines are doing. Maybe an indication that this could use some more documentation. (Or that I need sleep 😜)